From 1c55c7684f937cdc7d8246747d94556e4e133058 Mon Sep 17 00:00:00 2001
From: Bin Jiang <bin.jiang@windriver.com>
Date: Tue, 18 Feb 2014 15:17:09 +0800
Subject: [PATCH] uio: align uio memory region

The uio memory region should be at page boundaries in order for
user mmap() to work. The region size is rounded to a multiple of
PAGE_SIZE.

Signed-off-by: Bin Jiang <bin.jiang@windriver.com>
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
Signed-off-by: Weiwei Wang <weiwei.wang@windriver.com>

diff --git a/drivers/uio/uio.c b/drivers/uio/uio.c
index 4e13d213b2ca..61b9f0f87197 100644
--- a/drivers/uio/uio.c
+++ b/drivers/uio/uio.c
@@ -832,6 +832,22 @@ static int uio_request_irq(struct uio_info *info, struct uio_device *idev)
 				    info->irq_flags, info->name, idev);
 }
 
+static void uio_mem_region_align(struct uio_info *info)
+{
+	int mi;
+	struct uio_mem *mem;
+
+	for (mi = 0; mi < MAX_UIO_MAPS; mi++) {
+		mem = &info->mem[mi];
+		if (mem->size == 0)
+			break;
+		if (mem->memtype == UIO_MEM_PHYS) {
+			mem->addr = mem->addr & PAGE_MASK;
+			mem->size = PAGE_ALIGN(mem->size);
+		}
+	}
+}
+
 /**
  * uio_register_device - register a new userspace IO device
  * @owner:	module that creates the new device
@@ -850,6 +866,8 @@ int __uio_register_device(struct module *owner,
 	if (!parent || !info || !info->name || !info->version)
 		return -EINVAL;
 
+	uio_mem_region_align(info);
+
 	info->uio_dev = NULL;
 
 	idev = devm_kzalloc(parent, sizeof(*idev), GFP_KERNEL);
-- 
2.1.0

